source/common/http/conn_pool_base.cc

```diff
@@ -1,40 +1,292 @@
+      transitionActiveClientState(client, ActiveClient::State::DRAINING);
+    } else if (client.codec_client_->numActiveRequests() >= client.concurrent_request_limit_) {
+      transitionActiveClientState(client, ActiveClient::State::BUSY);
+    }
+
+    host_->stats().rq_total_.inc();
+    host_->stats().rq_active_.inc();
+    host_->cluster().stats().upstream_rq_total_.inc();
+    host_->cluster().stats().upstream_rq_active_.inc();
+    host_->cluster().resourceManager(priority_).requests().inc();
+    callbacks.onPoolReady(new_encoder, client.real_host_description_,
+                          client.codec_client_->streamInfo());
+  }
 }
 
-void ConnPoolImplBase::ActiveClient::disarmConnectTimeout() {
-  if (connect_timer_) {
-    connect_timer_->disableTimer();
-    connect_timer_.reset();
+void ConnPoolImplBase::onRequestClosed(ActiveClient& client, bool delay_attaching_request) {
+  ENVOY_CONN_LOG(debug, "destroying stream: {} remaining", *client.codec_client_,
+                 client.codec_client_->numActiveRequests());
+  host_->stats().rq_active_.dec();
+  host_->cluster().stats().upstream_rq_active_.dec();
+  host_->cluster().resourceManager(priority_).requests().dec();
+  if (client.state_ == ActiveClient::State::DRAINING &&
+      client.codec_client_->numActiveRequests() == 0) {
+    // Close out the draining client if we no longer have active requests.
+    client.codec_client_->close();
+  } else if (client.state_ == ActiveClient::State::BUSY) {
+    // A request was just ended, so we should be below the limit now.
+    ASSERT(client.codec_client_->numActiveRequests() < client.concurrent_request_limit_);
+
+    transitionActiveClientState(client, ActiveClient::State::READY);
+    if (!delay_attaching_request) {
+      onUpstreamReady();
+    }
   }
 }
 
-ConnPoolImplBase::ActiveClient::ConnectionState ConnPoolImplBase::ActiveClient::connectionState() {
-  // We don't track any failure state, as the client should be deferred destroyed once a failure
-  // event is handled.
-  if (connect_timer_) {
-    return Connecting;
+ConnectionPool::Cancellable* ConnPoolImplBase::newStream(Http::StreamDecoder& response_decoder,
+                                                         ConnectionPool::Callbacks& callbacks) {
+  if (!ready_clients_.empty()) {
+    ActiveClient& client = *ready_clients_.front();
+    ENVOY_CONN_LOG(debug, "using existing connection", *client.codec_client_);
+    attachRequestToClient(client, response_decoder, callbacks);
+    return nullptr;
   }
 
-  return Connected;
+  if (host_->cluster().resourceManager(priority_).pendingRequests().canCreate()) {
+    const bool can_create_connection =
+        host_->cluster().resourceManager(priority_).connections().canCreate();
+    if (!can_create_connection) {
+      host_->cluster().stats().upstream_cx_overflow_.inc();
+    }
+
+    // If we have no connections at all, make one no matter what so we don't starve.
```

Comment:
It's a bit non-obvious why we always create connections but we don't always create requests in the else block before.  mind adding a bit more detail to the comments?

Reply:
I know where this code came from before, but I'm wondering with this new code if this is still required? Meaning, how is it possible that we would have zero connections not be able to make a connection? That seems like a config error that should be caught elsewhere. I think we can probably get rid of this now?

Reply:
I don't know what this was for originally, but one case that comes to mind is if we hit the connection circuit breaker due to many active/draining connections for other pool members, and are unable to create a single connection to this pool member. I'm not sure if that case requires any special case or not. I can delete this if you're sure it's not needed.

Reply:
> but one case that comes to mind is if we hit the connection circuit breaker due to many active/draining connections for other pool members, and are unable to create a single connection to this pool member.

Hmm right, I think that might still be needed. Can you add more comments here and we can decide later whether we still need this or not?